stable_baselines/common/segment_tree.py

```diff
@@ -61,29 +62,42 @@ def reduce(self, start=0, end=None):
         return self._reduce_helper(start, end, 1, 0, self._capacity - 1)
 
     def __setitem__(self, idx, val):
-        # index of the leaf
-        idx += self._capacity
-        self._value[idx] = val
-        idx //= 2
-        while idx >= 1:
-            self._value[idx] = self._operation(
-                self._value[2 * idx],
-                self._value[2 * idx + 1]
-            )
-            idx //= 2
+        # indexes of the leaf
+        idxs = idx + self._capacity
+        self._value[idxs] = val
+        if isinstance(idxs, np.ndarray):
+            # rebuild the tree (should speed things up due to vectorization)
+            idxs = np.arange(self._capacity)+self._capacity
+            idxs //= 2
+            # by construction, all indexes reach 0 at the same time
+            while np.any(idxs >= 1):
+                self._value[idxs] = self._operation(
+                    self._value[2 * idxs],
+                    self._value[2 * idxs + 1]
+                )
+                idxs //= 2
+        else:
+            idxs //= 2
+            while idxs >= 1:
+                self._value[idxs] = self._operation(
+                    self._value[2 * idxs],
+                    self._value[2 * idxs + 1]
+                )
+                idxs //= 2
 
     def __getitem__(self, idx):
-        assert 0 <= idx < self._capacity
+        assert 0 <= np.max(idx) < self._capacity
```

Comment:
To match previous behavior this should check validity of all indices in `idx`, not just `max(idx)`. Something along lines of `assert np.all((0 <= idx) & (idx < self._capacity))`, if my memory serves me well.